### DISCUSSION TITLE
Root invoke reinitiate everytime machine transition states.

### DISCUSSION THREAD

**Discussion by dsngo** [statelyai/xstate#1819] [Q&A] [answered]
When the machine receives event to transition to child state, the root invoke reinitiate again. I do not know if this the desired behaviour of xstate?
Edit: I also find that the exit action fires every time as well. I do not really understand because since the root machine is supposed to be unchanged?
Working repo: https://codesandbox.io/s/nifty-yalow-vswjj?file=/src/index.js

**Comment by Andarist** [comment, marked as answer]
You are using an external transition that re-enters its source state (the one on which `TEST` transition is defined). So this looks like the desired behavior - you can make that transition internal though to change this. You can do it either by targeting `.inactive` instead of `inactive` or defining the transition like this:
```ts
on: { TEST: { target: 'inactive', internal: true } }
```

**Reply by dsngo** [reply]
Thank you.

**Reply by anthonyvia** [reply]
@Andarist I wanted to double check the intended behavior here. It still seems odd to me that under these circumstances the root state node undergoes a transition even though the state that is actually being transitioned is one of the states defined in the `states:` config. I guess it is just unintuitive to me that the event transition handler defined in the root config causes that root state node to undergo a transition in addition to the child state undergoing a transition.

In case anyone else comes across this issue, I have a working example that I forked from the OP's sandbox that is a bit easier to see this in play: https://codesandbox.io/s/anthony-xstate-behavior-5rxhz?file=/src/index.js. Just open the console, then click the toggle button in the app to see the console output. Update the `target`s for the `TOGGLE` event handlers if you want to play with the external/internal transitions to see how the invoked service gets restarted on an external transition.

**Reply by Andarist** [reply]
@anthonyvia I agree with you that this is surprising and I think there is some potential to change this in the next major version. See the discussion here: https://github.com/statelyai/xstate/discussions/1829 and pitch in there with this additional example on why this is surprising etc
